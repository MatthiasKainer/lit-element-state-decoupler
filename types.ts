export type LitLikeElement = {
    requestUpdate: () => void
    dispatchEvent: (e: Event) => boolean
    updateComplete: Promise<void>
}

export type UpdateableLitLikeElement = {
    updated(_?: any): void
}

type States = {
    index: number,
    count: number,
    states: InjectableState<any>[]
    reducers: Reduce<any>[]
    workflows: Workflow[]
}

export type DecoratedLitLikeElement = LitLikeElement & {
    __registered_states: States
}

export type Reducer<T> = (state: T) => {[action: string]: (payload?: any | unknown | undefined) => Promise<T>}

export type StateOptions = {
    updateDefault?: boolean
}

export type ReducerOptions = StateOptions & {
    dispatchEvent?: boolean
}
export interface State<T> {
    /**
     * @deprecated The method should not be used. Use `get` or `value` instead.
     */
    getState: () => T
    /**
     * @deprecated The method should not be used. Use `set` or `value` instead.
     */
    publish: (update: T) => void
    get: () => T
    set: (update: T) => Promise<void>
    value: T
    subscribe: (onChange: (state: T) => Promise<void>) => void
}
export interface InjectableState<T> extends State<T> {
    inject: (update: T) => void,
}

export type Reduce<T> = {
    get: () => T
    set: <T>(action: string, data?: T) => Promise<T>
    subscribe: (onChange: (action: string, state: T) => void) => void
    when: (action: string, onChange: (state: T) => void) => void
    dispatch: <T>(action: string, data?: T) => Promise<T>
}

export type WorkflowHistory = { type: string, args: unknown[] }

export type Workflow = {
    addActivity: (activity: string, data?: unknown) => Promise<void>
    addCompensation: (activity: string, data?: unknown) => void
    addSideeffect: (activity: string, sideeffect: (data?: unknown) => Promise<unknown>) => void
    projections: (key: string) => any
    after: (timeout: Date, unlessActivity: WorkflowHistory, execute: () => Promise<unknown>) => void
    compensate: () => Promise<unknown>
    history: () => WorkflowHistory[]
    plan: <T>(plan: {[entity: string]: () => Promise<T>}) => Promise<T | null>
}