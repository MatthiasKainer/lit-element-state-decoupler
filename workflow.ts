import { useReducer } from ".";
import { withWorkflow } from "./decorator";
import { LitLikeElement, Reduce, Reducer, Workflow, WorkflowHistory } from "./types"

type InnerReducers = { [key: string]: Reduce<any> }
type Compensation = { [key: string]: unknown[] }
type Sideeffects = { [key: string]: ((data?: unknown) => Promise<unknown>)[] }

export const useWorkflow = (element: LitLikeElement, reducers: { [key: string]: {reducer: Reducer<any>, initialState: unknown} }): Workflow => {
    const innerReducers = Object.entries(reducers).reduce((prev, [projection, {reducer, initialState}]) =>
        (prev[projection] = useReducer(element, reducer, initialState), prev)
        , {} as InnerReducers)
    const sideeffect: Sideeffects = {}
    const compensations: Compensation = {}
    const workflowHistory: WorkflowHistory[] = []

    const projections = (key: string) => {
        workflowHistory.push({
            type: "projections",
            args: [key]
        })
        return (innerReducers[key]) ? innerReducers[key].get() : undefined
    }

    const addActivity = async (activity: string, data?: unknown) => {
        workflowHistory.push({
            type: "addActivity",
            args: [activity, data]
        })
        await Promise.all(sideeffect[activity]?.map(effect => effect(data)) ?? [])
        for (const reducer of Object.values(innerReducers)) {
            await reducer.dispatch(activity, data)
        }
    }

    const addCompensation = (activity: string, data?: unknown) => {
        workflowHistory.push({
            type: "addCompensation",
            args: [activity, data]
        })
        compensations[activity] = [
            ...(compensations[activity] || []),
            data
        ];
    }

    const addSideeffect = (activity: string, effect: (data?: unknown) => Promise<unknown>) => {
        workflowHistory.push({
            type: "addSideeffect",
            args: [activity, effect]
        })
        sideeffect[activity] = [
            ...(sideeffect[activity] || []),
            effect
        ];
    }

    const compensate = async () => {
        workflowHistory.push({
            type: "compensate",
            args: []
        })
        for (const [activity, dataArguments] of Object.entries(compensations)) {
            for (const data of dataArguments) {
                for (const reducer of Object.values(innerReducers)) {
                    await reducer.dispatch(activity, data)
                }
            }
        }
    }

    const after = (timeout: Date, unlessActivity: WorkflowHistory, execute: () => Promise<unknown>) => {
        workflowHistory.push({
            type: "after",
            args: [timeout, unlessActivity, execute]
        })
        const compareWorkflow = (entry: WorkflowHistory, unless: WorkflowHistory) => {
            if (entry.type !== unless.type) return false;
            for (let arg = 0; arg < unless.args.length; arg++) {
                if (entry.args[arg] !== unless.args[arg]) return false;
            }
            return true;
        }
        const check = async () => {
            if (workflowHistory.some(entry => compareWorkflow(entry, unlessActivity))) {
                return;
            } else if (new Date(Date.now()) > timeout) {
                await execute()
            } else {
                setTimeout(check, 100)
            }
        }
        check()
    }

    const plan = async <T>(plan: {[entity: string]: () => Promise<T>}) => {
        for (const [entity, workflow] of Object.entries(plan)) {
            if (reducers[entity] &&
                JSON.stringify(projections(entity)) === JSON.stringify(reducers[entity].initialState)) {
                return await workflow()
            }
        }
        return plan[""] 
            ? await plan[""]() 
            : Promise.resolve(null)
    }

    return withWorkflow(element, {
        addActivity,
        addSideeffect,
        projections,
        addCompensation,
        compensate,
        after,
        plan,
        history: () => [...workflowHistory]
    })
}