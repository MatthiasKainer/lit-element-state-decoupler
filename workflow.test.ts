import { Reducer } from ".";
import { LitLikeElement, Workflow } from "./types"
import { useWorkflow } from "./workflow";

type Customer = {
    userName: string
}

type Contract = {
    contractNumber: number,
    userName: string
    created: Date
    ends: Date
}

type ErrorCodes = "NOT_FOUND" | "CONFLICT"
class Rejection<T> {
    code: ErrorCodes
    message: string
    state: T

    constructor(code: ErrorCodes, message: string, state: T) {
        this.code = code;
        this.message = message
        this.state = state
    }

    public toString() {
        return `[${this.code}] ${this.message}`;
    }
}

const notFound = <T>(who: string, state: T) => (console.log(`${who} not found`, state), new Rejection("NOT_FOUND", `${who} not found`, state));
const conflict = <T>(who: string, state: T) => (console.log(`${who} already exists`, state), new Rejection("CONFLICT", `${who} already exists`, state));

const customersReducer: Reducer<Customer[]> = (state: Customer[]) => ({
    createCustomer: (userName: string) => {
        return new Promise((resolve, reject) => {
            if (state.some(customer => customer.userName === userName)) return reject(conflict("User", state));
            return resolve([...state, { userName }] as Customer[])
        });
    },
    changeCustomerName: ({ oldName, userName }: { oldName: string, userName: string }) => new Promise((resolve, reject) => {
        const index = state.findIndex((customer) => customer.userName === oldName);
        if (index < 0) return reject(notFound("User", state));
        state[index] = { ...state[index], userName }
        return resolve([...state] as Customer[])
    }),
    deleteCustomer: (userName: string) => new Promise((resolve, reject) => {
        const index = state.findIndex((customer) => customer.userName === userName);
        if (index < 0) return (reject(notFound("User", state)));
        state.splice(index, 1)
        return resolve([...state] as Customer[])
    })
})

const contractReducer: Reducer<Contract[]> = (state: Contract[]) => ({
    createContract: ({ contractNumber, userName, runtime }: { contractNumber: number, userName: string, runtime: number }) => new Promise((resolve, reject) => {
        if (state.some(contract => contract.contractNumber === contractNumber)) return reject(conflict("Contract", state));
        const created = new Date(Date.now())
        // contract always starts at the beginning of the month
        created.setDate(1)
        const ends = new Date(created)
        ends.setMonth(ends.getMonth() + runtime)
        return resolve([
            ...state,
            { contractNumber, userName: userName, created, ends }
        ])
    }),
    changeCustomerName: ({ oldName, userName }: { oldName: string, userName: string }) => new Promise((resolve) =>
        resolve(state.map(contract =>
            (contract.userName === oldName) ? { ...contract, userName } : contract
        ))
    ),
    removeContract: (contractNumber: number) => new Promise((resolve, reject) => {
        const index = state.findIndex(contract => contract.contractNumber === contractNumber);
        if (index < 0) return reject(notFound("Contract", state));
        state.splice(index, 1)
        return resolve([...state])
    })
})

describe("Given I have a workflow", () => {

    let workflow: Workflow;


    beforeEach(() => {
        jest.resetAllMocks()
        const litElement = {
            requestUpdate: jest.fn(),
            dispatchEvent: jest.fn()
        } as unknown as LitLikeElement
        workflow = useWorkflow(litElement, {
            customers: { reducer: customersReducer, initialState: [] },
            contracts: { reducer: contractReducer, initialState: [] }
        })
    })

    it("returns undefined for non-existing projections (probably a bad idea, let's think about something better)", () => {
        expect(workflow.projections("made up")).toEqual(undefined)
    })

    it("executes every step when all go well", async () => {
        let userName = "Mustermann"
        await workflow.addActivity("createCustomer", userName);
        expect(workflow.projections("customers")).toEqual([{ userName }])
        expect(workflow.projections("contracts")).toEqual([])

        const contractNumber = 1
        await workflow.addActivity("createContract", { userName, contractNumber, runtime: 3 });
        expect(workflow.projections("customers")).toEqual([{ userName }])
        expect(workflow.projections("contracts")).toEqual([{ contractNumber, "created": expect.any(Date), "ends": expect.any(Date), userName }])

        userName = "Klaus"
        await workflow.addActivity("changeCustomerName", { oldName: "Mustermann", userName })
        expect(workflow.projections("customers")).toEqual([{ userName }])
        expect(workflow.projections("contracts")).toEqual([{ contractNumber, "created": expect.any(Date), "ends": expect.any(Date), userName }])

        workflow.addSideeffect("deleteCustomer", async (userName) =>
            await Promise.all((workflow.projections("contracts") as Contract[])
                .filter(contract => contract.userName === userName)
                .map(({ contractNumber }) => workflow.addActivity("removeContract", contractNumber)))
        )
        await workflow.addActivity("deleteCustomer", userName)
        expect(workflow.projections("customers")).toEqual([])
        expect(workflow.projections("contracts")).toEqual([])

        expect(workflow.history()).toMatchSnapshot()
    })

    it("compensates if a step fails", async () => {
        let userName = "Mustermann"
        const contractNumber = 1
        await workflow.addActivity("createCustomer", userName);
        workflow.addCompensation("deleteCustomer", userName);
        await workflow.addActivity("createContract", { userName, contractNumber, runtime: 3 });
        workflow.addCompensation("removeContract", contractNumber);

        expect(workflow.projections("customers")).toEqual([{ userName }])
        expect(workflow.projections("contracts")).toEqual([{ contractNumber, "created": expect.any(Date), "ends": expect.any(Date), userName }])

        // we just want to avoid it gets triggered right away, just for readability
        const days = (_: number) => {
            const future = new Date();
            future.setMilliseconds(future.getMilliseconds() + 100);
            return future
        }

        workflow.after(days(1), { type: "confirmedEmail", args: [userName] }, async () => await workflow.compensate())
        await new Promise(resolve => setTimeout(resolve, 500))

        // compensation should have happenend
        expect(workflow.projections("customers")).toEqual([])
        expect(workflow.projections("contracts")).toEqual([])
        expect(workflow.history().some(entry => entry.type === "compensate")).toBeTruthy()

    })

    it("does not compensates if all step pass on a full history match", async () => {
        let userName = "Mustermann"
        const contractNumber = 1
        await workflow.addActivity("createCustomer", userName);
        workflow.addCompensation("deleteCustomer", userName);
        await workflow.addActivity("createContract", { userName, contractNumber, runtime: 3 });
        workflow.addCompensation("removeContract", contractNumber);

        expect(workflow.projections("customers")).toEqual([{ userName }])
        expect(workflow.projections("contracts")).toEqual([{ contractNumber, "created": expect.any(Date), "ends": expect.any(Date), userName }])

        // we want that to be now anyways, just for readability
        const days = (_: number) => new Date(0);

        // it should not matter if an activity has happened before or after in a workflow
        //  this way we won't have timing issues as after will be executed right away
        await workflow.addActivity("confirmedEmail", userName);

        workflow.after(days(1), { type: "addActivity", args: ["confirmedEmail", userName] }, async () => await workflow.compensate())

        // should not be componsated
        expect(workflow.projections("customers")).toEqual([{ userName }])
        expect(workflow.projections("contracts")).toEqual([{ contractNumber, "created": expect.any(Date), "ends": expect.any(Date), userName }])
        expect(workflow.history().some(entry => entry.type === "compensate")).toBeFalsy()

    })
    it("does not compensates if all step pass when checking for the start of a history entry", async () => {
        let userName = "Mustermann"
        const contractNumber = 1
        await workflow.addActivity("createCustomer", userName);
        workflow.addCompensation("deleteCustomer", userName);
        await workflow.addActivity("createContract", { userName, contractNumber, runtime: 3 });
        workflow.addCompensation("removeContract", contractNumber);

        expect(workflow.projections("customers")).toEqual([{ userName }])
        expect(workflow.projections("contracts")).toEqual([{ contractNumber, "created": expect.any(Date), "ends": expect.any(Date), userName }])

        // we want that to be now anyways, just for readability
        const days = (_: number) => new Date(0);

        // it should not matter if an activity has happened before or after in a workflow
        //  this way we won't have timing issues as after will be executed right away
        await workflow.addActivity("confirmedEmail", userName);

        workflow.after(days(1), { type: "addActivity", args: ["confirmedEmail"] }, async () => await workflow.compensate())

        // should not be componsated
        expect(workflow.projections("customers")).toEqual([{ userName }])
        expect(workflow.projections("contracts")).toEqual([{ contractNumber, "created": expect.any(Date), "ends": expect.any(Date), userName }])
        expect(workflow.history().some(entry => entry.type === "compensate")).toBeFalsy()

    })
})

describe("Given I plan a workflow", () => {
    let workflow: Workflow;


    beforeEach(() => {
        jest.resetAllMocks()
        const litElement = {
            requestUpdate: jest.fn(),
            dispatchEvent: jest.fn()
        } as unknown as LitLikeElement
        workflow = useWorkflow(litElement, {
            customers: { reducer: customersReducer, initialState: [] },
            contracts: { reducer: contractReducer, initialState: [] }
        })
    })

    it("executes the plan in the correct order", async () => {
        const mockTrace = jest.fn();
        const userName = "name"
        const contractNumber = 1
        const plan = {
            customers: async () => {
                await workflow.addActivity("createCustomer", userName);
                mockTrace("customer")
                return "customer"
            },
            contracts: async () => {
                await workflow.addActivity("createContract", { userName, contractNumber, runtime: 3 });
                mockTrace("contracts")
                return "contracts"
            },
            "" : async () => {
                mockTrace("done")
                return "done"
            }
        }
        let result = await workflow.plan(plan)

        await new Promise(process.nextTick)

        expect(mockTrace.mock.calls[0][0]).toBe("customer")
        result = await workflow.plan(plan)
        expect(mockTrace.mock.calls[1][0]).toBe("contracts")
        result = await workflow.plan(plan)
        expect(result).toBe("done")
    })

    it("finishes a plan with an empty last result if none given", async () => {
        const mockTrace = jest.fn();
        const plan = {
            customers: async () => {
                await workflow.addActivity("createCustomer", "name");
                mockTrace("customer")
                return "customer"
            },
            madeUp: async () => "made up!"
        }
        let result = await workflow.plan(plan)

        await new Promise(process.nextTick)

        expect(mockTrace.mock.calls[0][0]).toBe("customer")
        result = await workflow.plan(plan)
        expect(result).toBe(null)
    })
})