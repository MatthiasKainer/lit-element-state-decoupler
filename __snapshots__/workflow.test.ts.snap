// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given I have a workflow executes every step when all go well 1`] = `
Array [
  Object {
    "args": Array [
      "createCustomer",
      "Mustermann",
    ],
    "type": "addActivity",
  },
  Object {
    "args": Array [
      "customers",
    ],
    "type": "projections",
  },
  Object {
    "args": Array [
      "contracts",
    ],
    "type": "projections",
  },
  Object {
    "args": Array [
      "createContract",
      Object {
        "contractNumber": 1,
        "runtime": 3,
        "userName": "Mustermann",
      },
    ],
    "type": "addActivity",
  },
  Object {
    "args": Array [
      "customers",
    ],
    "type": "projections",
  },
  Object {
    "args": Array [
      "contracts",
    ],
    "type": "projections",
  },
  Object {
    "args": Array [
      "changeCustomerName",
      Object {
        "oldName": "Mustermann",
        "userName": "Klaus",
      },
    ],
    "type": "addActivity",
  },
  Object {
    "args": Array [
      "customers",
    ],
    "type": "projections",
  },
  Object {
    "args": Array [
      "contracts",
    ],
    "type": "projections",
  },
  Object {
    "args": Array [
      "deleteCustomer",
      [Function],
    ],
    "type": "addSideeffect",
  },
  Object {
    "args": Array [
      "deleteCustomer",
      "Klaus",
    ],
    "type": "addActivity",
  },
  Object {
    "args": Array [
      "contracts",
    ],
    "type": "projections",
  },
  Object {
    "args": Array [
      "removeContract",
      1,
    ],
    "type": "addActivity",
  },
  Object {
    "args": Array [
      "customers",
    ],
    "type": "projections",
  },
  Object {
    "args": Array [
      "contracts",
    ],
    "type": "projections",
  },
]
`;
